### ISSUE TITLE
Some apps are always shown as outdated even though they're already the latest version (CFBundleShortVersionString vs. CFBundleVersion)

### ISSUE THREAD

**Issue by luispuerto** [mas-cli/mas#384]
## Your Environment

- `mas version`: 1.8.2
- macOS version (`system_profiler SPSoftwareDataType -detailLevel mini`): 11.3.1 (20E241)

### mas Install Method

- [x] `brew install mas` (homebrew-core)
- [ ] [mas-cli/tap](https://github.com/mas-cli/homebrew-tap)
- [ ] `.pkg` installer from [releases](https://github.com/mas-cli/mas/releases)
- [ ] Built from source
  - Fork/branch: ? (e.g. mas-cli/master)
  - Xcode version: 10.?

## Describe the Bug

Ninja Cookie gets updated from 0.7 to 0.7.1, but if I run `mas upgrade` it updates again because `mas outdated` still show it as outdated. On other words, just seem that or mas is not able to read the version of the app properly, or the app doesn't states its version properly. I think it's the latter one. 

## To Reproduce

1. Install Ninja Cookie 1535219336
2. `mas outdated` or `mas upgrade`
3. See that Ninja Cookie upgrades every time

## Expected Behavior

Detect the version properly? 

## Actual Behavior

Upgrade, upgrade, upgrade

## Additional Context

Perhaps it would be great to have a pin command or ignore this version or something.

**Comment by lougreenwood** [comment]
I also get this for `1Blocker 3.16` and `Endel 2.3.264`.

Not sure if these should go in separate issues as application specific or if it's a general `mas` bug?

**Comment by luispuerto** [comment]
@lougreenwood I don't know either and at least in the case of Ninja Cookie the short version is just 0.7, so perhaps that the confusion.

**Comment by phatblat** [comment]
Yep, I get the same, even immediately after installing the app.

```shell
↪️ mas search "ninja cookie"
  1535219336  Ninja Cookie (0.7.1)

↪️  mas info 1535219336   
Ninja Cookie 0.7.1 [0.0]
By: Theo GOUDOUT
Released: 2021-05-14
Minimum OS: 10.14
Size: 4 MB
From: https://apps.apple.com/us/app/ninja-cookie/id1535219336?mt=12&uo=4

↪️  mas purchase 1535219336
==> Downloading Ninja Cookie
==> Installed Ninja Cookie

↪️  mas outdated
1535219336 Ninja Cookie (0.7 -> 0.7.1)
```

**Comment by phatblat** [comment]
Sorry. There have been a handful of annoying issues like this since we fixed the `outdated` and `upgrade` commands.

Noticed that both the version numbers in the `outdated` output are listed in the app's "About" window and `Info.plist` file.

<img width="396" alt="Screen Shot 2021-05-19 at 10 46 38 PM" src="https://user-images.githubusercontent.com/28851/118921464-13919500-b8ed-11eb-87fd-94e6f795cccd.png">

```xml
<key>CFBundleShortVersionString</key>
<string>0.7</string>
<key>CFBundleVersion</key>
<string>0.7.1</string>
```

**Comment by chris-araman** [comment]
I recently fixed the case where the bundle version is newer than the App Store version. This is the opposite case. We don't have a way to get the bundle version from the App Store without downloading the app again every time we want to check for the latest release.

We could heuristically parse both `CFBundleShortVersionString` and `CFBundleVersion`, and if they're both parseable as Semantic Versions, use the greater of the two as the "installed" version to be compared against the App Store version. It'll catch cases like this one, but there are probably still more edge and corner cases.

**Comment by PaulDoesDev** [comment]
I'm still seeing this myself with `1Blocker 3.16.2`

**Comment by chris-araman** [comment]
> We could heuristically parse both `CFBundleShortVersionString` and `CFBundleVersion`, and if they're both parseable as Semantic Versions, use the greater of the two as the "installed" version to be compared against the App Store version.

It seems this plan isn't going to work, as `CFBundleVersion` seems to have no relation to `CFBundleShortVersionString` for many apps.

As things are today, using only `CKSoftwareProduct.bundleVersion`, which is generally expected to match `CFBundleShortVersionString`:
```
1489275350  Crossy Castle             (5.1)
1480068668  Messenger                 (97.11.116)
640199958   Developer                 (9.2.2)
1376509945  Populus Run               (1.3)
409183694   Keynote                   (11.1)
784801555   Microsoft OneNote         (16.51)
823766827   OneDrive                  (21.129.0627)
462058435   Microsoft Excel           (16.51)
1142578753  OmniGraffle               (7.18.5)
1482454543  Twitter                   (8.75)
1462114288  Grammarly for Safari      (9.36)
1496833156  Playgrounds               (3.4.1)
682658836   GarageBand                (10.4.3)
462062816   Microsoft PowerPoint      (16.51)
1206020918  Battery Indicator         (2.9.1)
1289197285  MindNode                  (2021.2.2)
1153157709  Speedtest                 (1.20)
409203825   Numbers                   (11.1)
497799835   Xcode                     (12.5.1)
1464883939  Frogger in Toy Town       (1.5.2)
409201541   Pages                     (11.1)
1448916662  Step Two                  (3.1.1)
1475387142  Tailscale                 (1.10.1)
408981434   iMovie                    (10.2.5)
985367838   Microsoft Outlook         (16.51)
1295203466  Microsoft Remote Desktop  (10.6.7)
1450874784  Transporter               (1.2.2)
803453959   Slack                     (4.18.0)
1470584107  Dato                      (3.0.1)
462054704   Microsoft Word            (16.51)
1263070803  Lungo                     (1.9.0)
```

However,  using the greater of `CKSoftwareProduct.bundleVersion`, `CFBundleShortVersionString`, and `CFBundleVersion`, we see that `CFBundleVersion` is frequently unrelated to the former two:
```
1489275350  Crossy Castle             (7740.0.0)
1480068668  Messenger                 (283133794.0.0)
640199958   Developer                 (922.3.2)
1376509945  Populus Run               (210630606.0.0)
409183694   Keynote                   (7031.0.102)
784801555   Microsoft OneNote         (16.51.21071101)
823766827   OneDrive                  (21129.627.2)
462058435   Microsoft Excel           (16.51.21071101)
1142578753  OmniGraffle               (204.16.0)
1482454543  Twitter                   (8.75.0)
1462114288  Grammarly for Safari      (16.0.302203)
1496833156  Playgrounds               (1302.34.0)
682658836   GarageBand                (5433.0.0)
462062816   Microsoft PowerPoint      (16.51.21071101)
1206020918  Battery Indicator         (33.0.0)
1289197285  MindNode                  (2021.2.2)
1153157709  Speedtest                 (1.20.4)
409203825   Numbers                   (7031.0.102)
497799835   Xcode                     (18212.0.0)
1464883939  Frogger in Toy Town       (341.0.0)
409201541   Pages                     (7031.0.102)
1448916662  Step Two                  (23.0.0)
1475387142  Tailscale                 (101.10.1)
408981434   iMovie                    (400096.0.0)
985367838   Microsoft Outlook         (16.51.21071101)
1295203466  Microsoft Remote Desktop  (1896.0.0)
1450874784  Transporter               (1256.0.0)
803453959   Slack                     (6748.0.0)
1470584107  Dato                      (58.0.0)
462054704   Microsoft Word            (16.51.21071101)
1263070803  Lungo                     (18.0.0)
```

Unfortunately, more reports of the problem aren't helping us reach a solution. We understand the cause of the problem, but we need a viable proposal for how to address it in order to move forward.

Perhaps there's some investigation to be done into the private CommerceKit and StoreFoundation that might shed some light on this sort of versioning discrepancy? Maybe some magic to `CKSoftwareProduct.versionIdentifier`, `expectedBundleVersion`, or `expectedStoreVersion`? 🤷🏻‍♂️✨

**Comment by melgu** [comment]
I have the same issues with:
```
1515959873 Theu2007Survivalists (1.0.747221 -> 1.4)
1513822917 Baldo The Guardian Owls (1.1.1 -> 1.1.4)
1538650027 Alto's Odyssey The Lost City (1.1.1 -> 1.1.2)
```
Running e.g. `mas info 1538650027 ` shows the latest version as 1.1.2, but the English as well as the German App Store pages show the latest version as 1.1.1.

**Comment by aumuell** [comment]
I get a current version of 1Blocker reported as outdated:
`1365531024 1Blocker (4.1 -> 5.1)`

I would like to highlight this, as I didn't see this mentioned yet:
On the German Mac App Store, 4.1 is the current version of 1Blocker, but 5.1 is the current version of 1Blocker in the iOS store.

I'm on 12.0.1 on Apple Silicon, so also some apps from the iOS store are installable, but not 1Blocker. `mas` reports version 1.8.5.

**Comment by melgu** [comment]
I have the same problem.
```
1474220585 StrandedSails (2.0.4 -> 2.1.0)
1461650987 Crouton (2021.9.2 -> 2021.10)
1515959873 Theu2007Survivalists (1.0.767225 -> 1.4)
```
All of those already are on the latest version available for macOS.

**Comment by chris-araman** [comment]
I've summarized this in the [Known Issues](https://github.com/mas-cli/mas#%EF%B8%8F-known-issues) section of the README.

Additional reports of apps with mismatched app bundle versions are unfortunately not helpful at this point. Pull requests are welcome!

**Comment by ChrisChinchilla** [comment]
Out of interest (and yes I have my own persistent list of apps that are always getting updated), I also use CleanMy Mac and it's updater list generally has the correct versions for the same two apps, could be worth asking them how they get updated version lists?

**Comment by TheDen** [comment]
I had this issue with Zentuner, where it would upgrade every time
```bash
$ mas outdated
1550640624 ZenTuner (0.2.0 -> 0.2.2)
```

The workaround was to edit `/Applications/ZenTuner.app/Contents/Info.plist`

```XML
...
  <key>CFBundleShortVersionString</key>
  <string>0.2.0</string>
...
```

and and  the value to `0.2.2`. This isn't recommended, but useful if you have `mas upgrade` running on schedule and don't want to reinstall certain applications every time

**Comment by nicerloop** [comment]
See Feature Request https://github.com/mas-cli/mas/issues/468

**Comment by koalalorenzo** [comment]
Same with mas 1.8.6

**Comment by rgoldberg** [comment]
@koalalorenzo For which app IDs and app versions is this occurring for you? What version of macOS? What country do you use for the App Store?

Thanks.

**Comment by rgoldberg** [comment]
@koalalorenzo Thanks for the info. If you can, please don't update/upgrade these apps so we can diagnose & fix the issue(s).

Can you install from the `main` branch of mas (either via nix or `brew install --HEAD mas`; if `mas` is already by brew, please uninstall it first via `brew uninstall mas`)? I've just merged a ton of fixes, but we won't be able to release them for a bit, so this would help determine if you issues have already been fixed.

If the code in `main` doesn't fix your issue:

- Are you Intel or Apple Silicon?
- What version of each of these apps is actually installed? You can get this from each app by opening the app, then opening the "<app-name>" menu -> "About <app-name>". Please copy all version strings from each dialog.
- What version of each of these app does the App Store GUI say is installed? Does the App Store GUI report that any updates are available? If so, what are the reported versions of those updates?

I know nothing about nix; I'm not sure if nix might cause any issues.

**Comment by koalalorenzo** [comment]
@rgoldberg 
> I've just merged a ton of fixes in `main`, but we won't be able to release them for a bit. To see if those fixes fix your problem, can you install from the `main` branch of mas (either via nix or `brew install --HEAD mas`; if `mas` was already installed by brew, please uninstall it first via `brew uninstall mas`)?

As I wrote, I am using Nix, not brew. I tried to install it using brew and the main branch HEAD commit, not much success over some dependencies requiring different versions of macOS 😅

![Screenshot 2024-10-15 at 15 30 27](https://github.com/user-attachments/assets/5b4ad4a4-b7db-4b64-afc0-21906f18aa4d)

I am on Apple Sillicon, I have this issue with both M1 Pro and M3 Pro.
The versions of the apps I have installed, is the version reported to the left (the outdated version according to mas) and if I open the App Store, it does not ask to update, and the version reported as the latest one is the version currently installed.

**Comment by koalalorenzo** [comment]
@rgoldberg 
> I've just merged a ton of fixes in `main`, but we won't be able to release them for a bit. To see if those fixes fix your problem, can you install from the `main` branch of mas (either via nix or `brew install --HEAD mas`; if `mas` was already installed by brew, please uninstall it first via `brew uninstall mas`)?

As I wrote, I am using Nix, not brew. I tried to install it using brew and the main branch HEAD commit, not much success over some dependencies requiring different versions of macOS 😅

![Screenshot 2024-10-15 at 15 32 59](https://github.com/user-attachments/assets/5a0c7620-e847-40c4-91cc-1be186c95d88)

I am on Apple Sillicon, I have this issue with both M1 Pro and M3 Pro.
The versions of the apps I have installed, is the version reported to the left (the outdated version according to mas) and if I open the App Store, it does not ask to update, and the version reported as the latest one is the version currently installed.

**Comment by rgoldberg** [comment]
@koalalorenzo I don't get those mismatched version errors when I build, and we don't get those errors on the GitHub test servers.

What version of macOS & Xcode did you get those errors on?

4 of your outdated apps all seem to be spurious confusions between the macOS & the iOS versions; the problem should be fixed in `main` via PR #496.

1 of your outdated apps, Slack, does appear to be truly outdated. Try updating it. If that works, then you should be fine once the new version of mas is released.

In the meantime, if you want the current version from `main`, we can try to fix the testing macOS versions errors, or you can see if you can disable testing from the mas formula build, either via command line options or via editing the formula itself. I don't know if that can be done in nix, or if you can pass a binary that you've built elsewhere to nix.

**Comment by rgoldberg** [comment]
@koalalorenzo I fixed the testing macOS version errors in `main`. Can you try `brew install --HEAD mas` to install the current unreleased `main` code, then let me know if everything works fine for you?

**Comment by tdimeco** [comment]
Hi! I have just tested the HEAD version, it seems to work well:

```
➜  mas outdated
360593530 Notability (14.9.17 -> 14.9.21)
1365531024 1Blocker (4.6 -> 5.8)
➜  brew uninstall mas             
...
➜  brew install --HEAD mas
...
➜  mas outdated
1444636541 Photomator (3.4.1 -> 3.4.2)
➜  ~ 
```

`mas` previously reported Notability and 1Blocker but not Photomator (which is the only one that actually needs to be updated). After the update, this is the expected result. 🎉 Thanks!

**Comment by rgoldberg** [comment]
Closing because most of the issues reported here seem to be instances of the problem described in #491.

The few other issues reported here seem to be instances of the problem described in #540.

Fixes for these issues are in `main` waiting to be released in the next version.

There is no evidence that the problems noted here stem from a mismatch between CFBundleShortVersionString & CFBundleVersion.

If anyone experiences any problems with incorrect `mas outdated` or `mas upgrade` results with the code in `main`, or with any stable mas releases > 1.8.6, please open a new issue. I will try to get @phatblat update the beta releases with these & other fixes soon, but, currently, the beta releases suffer from the same problems as 1.8.6.

**Comment by rgoldberg** [comment]
Closing because most of the issues reported here seem to be duplicates of the problem described in #491.

The few other issues reported here seem to be duplicates of the problem described in #540.

Fixes for these issues are in `main` waiting to be released in the next version (either 1.8.7, 1.9.0, or 2.0.0, whatever we decide to call it).

There is no real evidence that the problems noted here stem from a mismatch between CFBundleShortVersionString & CFBundleVersion.

If anyone experiences any problems with incorrect `mas outdated` or `mas upgrade` results with the code in `main`, or with any stable mas releases > 1.8.6, please open a new issue. I will try to get @phatblat update the beta releases with these & other fixes soon, but, currently, the beta releases suffer from the same problems as 1.8.6.
